spec/shrinkwrap-and-lock.md

```diff
@@ -0,0 +1,273 @@
+shrinkwrap has never been specified.
+
+## Managing the two files
+
+If at any time both `package-lock.json` and an `npm-shrinkwrap.json` exist
+then the former should be unlinked with a warning the latter used. This should happen on any command
+that would interact with them:
+
+* `npm install`
+* `npm update`
+* `npm rm`
+* `npm shrinkwrap`
+
+## File Format
+
+`package-lock.json` and `npm-shrinkwrap.json` are JSON documents described
+below.  The difference between a `package-lock.json` and an
+`npm-shrinkwrap.json` is that the former is never published. The `npm shrinkwrap` command
+renames the `package-lock.json` (or if missing, generates one.)
+
+### name
+
+The name of the package this is a shrinkwrap for. This must match what's in `package.json`.
+
+### version
+
+The version of the package this is a shrinkwrap for. This must match what's in `package.json`.
+
+### created-with *(new)*
+
+The `name@version` of the creator of this shrinkwrap.  The name part must be
+a valid npm package name.  The version part must be a valid semver version
+string. For example: `npm@5.0.0`
+
+This is non-normative and intended to help in isolating the cause of errors.
+
+### shrinkwrap-version *(new)*
+
+An integer version, starting at `1` with the version number of this document
+whose semantics were used when generating this `npm-shrinkwrap.json`.
+
+### package-integrity *(new)*
+
+A hash of the `package.json` that was in use when this shrinkwrap was
+created.
+
+### dependencies
+
+A mapping of package name to dependency object.  Dependency objects have the
+following properties:
+
+#### version *(changed)*
+
+This is a specifier that uniquely identifies this package and should be
+usable in fetching a new copy of it.
+
+* bundled dependencies: Regardless of source, this is a version number that is purely for informational purposes.
+* registry sources: This is a version number. (eg, `1.2.3`)
+* git sources: This is a git specifier with resolved committish. (eg, `git+https://example.com/foo/bar#115311855adb0789a0466714ed48a1499ffea97e`)
+* http tarball sources: This is the URL of the tarball. (eg, `https://example.com/example-1.3.0.tgz`)
+* local tarball sources: This is the file URL of the tarball. (eg `file:///opt/storage/example-1.3.0.tgz`)
+* local link sources: This is the file URL of the link. (eg `file:libs/our-module`)
+
+#### integrity *(new)*
+
+This is a [Standard Subresource
+Integrity](https://w3c.github.io/webappsec/specs/subresourceintegrity/) for
+this resource.
+
+* For bundled dependencies this is not included, regardless of source.
+* For registry sources, this is the `integrity` that the registry provided, or if one wasn't provided the SHA1 in `shasum`.
+* For git sources this is the specific commit hash we cloned from.
+* For remote tarball sources this is an integrity based on a SHA512 of
+  the file.
+* For local tarball sources: This is an integrity field based on the SHA512 of the file.
```

Comment:
Interesting that there will be potentially 3 different formats for integrity (SHA1, SHA512, the subresource Integrity format) - would this potentially increase complexity in consuming applications (such as npm, and potentially yarn)?